### PR TITLE
[MNT] Dockerized the app

### LINKED ISSUE #48: Dockerize the app
- [x] Create a DOCKERFILE for the app
- ~[ ] Set up a CI to automatically push the docker file~
- ~[ ] Ensure it aligns with the automation in place~

### PR DESCRIPTION
**Pull request by rmanaem**
- Closes #48 

## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+WORKDIR /app
+
+COPY package*.json .
+
+RUN npm ci
+
+COPY . .
+
+RUN npm run build
```
Comment (surchs) [inline Dockerfile]: could you explain how we get the ENV variables into this at runtime? i.e. when the container launches?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+EXPOSE 5173
+
+ENTRYPOINT npm run preview
```
Comment (surchs) [inline Dockerfile]: any reason to take this over the CMD? Are there any command suffixes it'd make sense to add here?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+COPY package*.json .
+
+RUN npm ci
+
+COPY . .
+
+RUN npm run build
+
+EXPOSE 5173
```
Comment (surchs) [inline Dockerfile:11]: why do we need this? `EXPOSE` doesn't "expose" the port: https://docs.docker.com/reference/dockerfile/#expose but is more of a doc. Is that the intent?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,15 @@
+FROM node:latest
```
Comment (surchs) [inline Dockerfile]: want to lock that down more with a version? I believe we use node 20 LTS. Would be good to have some consistency there

**Review by surchs — APPROVED** [review]
thanks @rmanaem, looks good.

Mind taking a look at the questions I posted - and then should be :cook:

File: Dockerfile (hunk 2 above)
Comment (surchs) [reply Dockerfile]: Just to add here: unless you figured out some special vite sauce I don't know, you'll probably have to add the bulls step as part of the entry point / cmd to ensure the runtime env variables get used here

File: Dockerfile (hunk 1 above)
Comment (rmanaem) [reply Dockerfile]: Similar to the Dockerfile in the old query tool using a `.env` file.

File: Dockerfile (hunk 4 above)
Comment (rmanaem) [reply Dockerfile]: Good call

File: Dockerfile (hunk 2 above)
Comment (rmanaem) [reply Dockerfile]: > any reason to take this over the CMD? Are there any command suffixes it'd make sense to add here?

I reckon we didn't want it to be overwritten.


> you'll probably have to add the bulls step as part of the entry point / cmd to ensure the runtime env variables get used here

build step? It's a couple of lines above it.

File: Dockerfile (hunk 3 above)
Comment (rmanaem) [reply Dockerfile:11]: Yea for documentation since vite uses a different port than Nuxt.

File: Dockerfile (hunk 2 above)
Comment (surchs) [reply Dockerfile]: Right, but this build step runs only once during image creation the way you set it up. I assume at that point you've made a bunch of static files that then get served with `npm run preview` whenever I actually launch a container from this image. So I'm wondering how you ensure that I can do this:

- docker build without providing the ENV variables (i.e. including not in my shell while building) and
- changing the ENV var every time I launch a new container from the image

In the old tool we had to build the app every time someone launched it: https://github.com/neurobagel/old-query-tool/blob/10ba7fce2b8903338386f466b81d63bcd24ee82e/Dockerfile#L15
So I'm wondering what you are doing here that makes this unnecessary

Comment (surchs) [reply Dockerfile]: to be clear: what we did with the old app wasn't great or smart. It was just a simple way around the problem of having the ENV vars baked into the static app. There are ways around this that don't require rebuilding, but afaik they involve some form of search-replace within the built app dist folder.

**Comment by rmanaem** [comment]
> ### ❌ Deploy Preview for _neurobagel-query-tool_ failed.
> Name 	Link
> 🔨 Latest commit 	[35d712e](https://github.com/neurobagel/query-tool/commit/35d712ea86a8f740c57fe1bac7d1ba815432c7a6)
> 🔍 Latest deploy log 	https://app.netlify.com/sites/neurobagel-query-tool/deploys/65e8c2851a71110008c3f119

@surchs I reckon we need to update the netlify build script.

File: Dockerfile (hunk 2 above)
Comment (surchs) [reply Dockerfile]: > I reckon we didn't want it to be overwritten.

That's `ENTRYPOINT`. `CMD` can be overwritten: https://docs.docker.com/reference/dockerfile/#entrypoint . But you can append commands to `ENTRYPOINT`

**Comment by surchs** [comment]
Should build now: https://deploy-preview-51--neurobagel-query-tool.netlify.app/?node=All

File: Dockerfile (hunk 2 above)
Comment (surchs) [reply Dockerfile]: aha aha, my bad. You and I are saying the same thing. All good on ENTRYPOINT then

**Review by surchs — APPROVED** [review]
:tada: :cook:

**Comment by surchs** [comment]
:rocket: PR was released in [`v0.1.0`](https://github.com/neurobagel/query-tool/releases/tag/v0.1.0) :rocket:

**Pull request merged**